### PR TITLE
PLANNER-2610 ScoreDirector SPI

### PR DESCRIPTION
**Pull request by triceo**

### CODE REVIEW

**Comment by triceo** [comment]
Jenkins run native tests.

**Comment by triceo** [comment]
jenkins run native tests

**Comment by triceo** [comment]
jenkins run native tests

**Comment by triceo** [comment]
Jenkins run native tests

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/ScoreDirectorFactoryFactory.java [hunk 1]
```diff
@@ -95,18 +77,29 @@ public ScoreDirectorFactoryFactory(ScoreDirectorFactoryConfig config) {
...
-        AbstractConstraintStreamScoreDirectorFactory<Solution_, Score_> constraintStreamScoreDirectorFactory =
-                buildConstraintStreamScoreDirectorFactory(solutionDescriptor);
-        IncrementalScoreDirectorFactory<Solution_, Score_> incrementalScoreDirectorFactory =
-                buildIncrementalScoreDirectorFactory(solutionDescriptor);
-        DroolsScoreDirectorFactory<Solution_, Score_> droolsScoreDirectorFactory = buildDroolsScoreDirectorFactory(
-                classLoader, solutionDescriptor);
+        // Load all known Score Director Factories via SPI.
+        ServiceLoader<ScoreDirectorFactoryProvider> scoreDirectorFactoryProviders =
+                ServiceLoader.load(ScoreDirectorFactoryProvider.class);
+        Map<ScoreDirectorType, AbstractScoreDirectorFactory<Solution_, Score_>> factories =
```
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/ScoreDirectorFactoryFactory.java:83]: Code style: Can we rename this variable to scoreDirectorFactoryMap?
It should end with Map, so it's clear that it's not an array (or list etc)

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/ScoreDirectorFactoryFactory.java [hunk 2]
```diff
@@ -95,18 +77,29 @@ public ScoreDirectorFactoryFactory(ScoreDirectorFactoryConfig config) {
...
-        EasyScoreDirectorFactory<Solution_, Score_> easyScoreDirectorFactory =
-                buildEasyScoreDirectorFactory(solutionDescriptor);
-        AbstractConstraintStreamScoreDirectorFactory<Solution_, Score_> constraintStreamScoreDirectorFactory =
-                buildConstraintStreamScoreDirectorFactory(solutionDescriptor);
-        IncrementalScoreDirectorFactory<Solution_, Score_> incrementalScoreDirectorFactory =
-                buildIncrementalScoreDirectorFactory(solutionDescriptor);
-        DroolsScoreDirectorFactory<Solution_, Score_> droolsScoreDirectorFactory = buildDroolsScoreDirectorFactory(
-                classLoader, solutionDescriptor);
+        // Load all known Score Director Factories via SPI.
+        ServiceLoader<ScoreDirectorFactoryProvider> scoreDirectorFactoryProviders =
```
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/ScoreDirectorFactoryFactory.java:81]: Code style: variable name is plural, but it's not an array
Can we rename it to scoreDirectorFactoryProviderLoader?

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/ScoreDirectorFactoryFactory.java [hunk 3]
```diff
@@ -95,18 +77,29 @@ public ScoreDirectorFactoryFactory(ScoreDirectorFactoryConfig config) {
...
+                factories.put(provider.getSupportedScoreDirectorType(), factory);
+            }
+        }
+        AbstractScoreDirectorFactory<Solution_, Score_> easyScoreDirectorFactory = factories.get(EASY);
+        AbstractScoreDirectorFactory<Solution_, Score_> constraintStreamScoreDirectorFactory =
+                factories.get(CONSTRAINT_STREAMS);
+        AbstractScoreDirectorFactory<Solution_, Score_> incrementalScoreDirectorFactory = factories.get(INCREMENTAL);
+        AbstractScoreDirectorFactory<Solution_, Score_> droolsScoreDirectorFactory = factories.get(DROOLS);
 
+        // Make sure only one SDF is truly available.
```
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/ScoreDirectorFactoryFactory.java:98]: Let's not abbreviate to SDF - I'd argue that reads easily for the author for the next few months, but not for other developers.

Do we need a comment at all? I think the method name pretty much self documents

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/ScoreDirectorFactoryFactory.java [hunk 4]
```diff
@@ -95,18 +77,29 @@ public ScoreDirectorFactoryFactory(ScoreDirectorFactoryConfig config) {
...
         checkMultipleScoreDirectorFactoryTypes(easyScoreDirectorFactory, constraintStreamScoreDirectorFactory,
                 incrementalScoreDirectorFactory, droolsScoreDirectorFactory);
```
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/ScoreDirectorFactoryFactory.java:100]: If this is an assertion, can we rename from "checkMultiple" to "assertOnlyOneScoreDirectorFactory"?
Also, I'd drop the "Type(s)" suffix, as it doesn't check types, it checks factory instances.

Comment (ge0ffrey) [reply optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/ScoreDirectorFactoryFactory.java:100]: Ok, old naming was bad, let's fix that :)

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/ScoreDirectorType.java [hunk 5]
```diff
@@ -0,0 +1,10 @@
+package org.optaplanner.core.impl.score.director;
+
+public enum ScoreDirectorType {
+
+    EASY,
+    INCREMENTAL,
+    DROOLS,
+    CONSTRAINT_STREAMS
```
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/ScoreDirectorType.java:8]: Order should be:
easy, constraint_streams, drools, incremental IIRC.
Definitely CS before DRL these days.

Comment (ge0ffrey) [reply optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/ScoreDirectorType.java:8]: Can we rename DROOLS to DRL for clarity?

File: optaplanner-quarkus-integration/optaplanner-quarkus/deployment/src/main/java/org/optaplanner/quarkus/deployment/OptaPlannerProcessor.java [hunk 6]
```diff
@@ -103,6 +107,21 @@ FeatureBuildItem feature() {
         return new FeatureBuildItem("optaplanner");
     }
 
+    @BuildStep
+    void registerScoreDirectorFactorySpi(BuildProducer<ServiceProviderBuildItem> services) {
+        String serviceName = ScoreDirectorFactoryProvider.class.getName();
+        String service = "META-INF/services/" + serviceName;
```
Comment (ge0ffrey) [inline optaplanner-quarkus-integration/optaplanner-quarkus/deployment/src/main/java/org/optaplanner/quarkus/deployment/OptaPlannerProcessor.java:113]: Doesn't Quarkus have an out-of-the-box way of doing this? Seems low level that we have to mimic this ServiceLoader stuff ourselves. Something to ask on quarkus-dev zulip?

File: optaplanner-core/src/main/resources/META-INF/services/org.optaplanner.core.impl.score.director.ScoreDirectorFactoryProvider [hunk 7]
```diff
@@ -0,0 +1,4 @@
+org.optaplanner.core.impl.score.director.drools.DroolsScoreDirectorFactoryProvider
+org.optaplanner.core.impl.score.director.easy.EasyScoreDirectorFactoryProvider
+org.optaplanner.core.impl.score.director.incremental.IncrementalScoreDirectorFactoryProvider
+org.optaplanner.core.impl.score.director.stream.ConstraintStreamsScoreDirectorFactoryProvider
```
Comment (ge0ffrey) [inline optaplanner-core/src/main/resources/META-INF/services/org.optaplanner.core.impl.score.director.ScoreDirectorFactoryProvider:4]: nitpick: order: easy, cs, incremental, drl

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/ScoreDirectorFactoryProvider.java [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+/**
+ * All {@link ScoreDirectorFactory} implementations must provide an implementation of this interface,
+ * as well as an entry in META-INF/services/org.optaplanner.core.impl.score.director.ScoreDirectorFactoryProvider file.
+ * This makes it available for discovery in {@link ScoreDirectorFactoryFactory} via {@link java.util.ServiceLoader}.
+ *
+ * @param <Solution_> the solution type, the class with the {@link PlanningSolution} annotation
+ * @param <Score_> the score type to go with the solution
+ */
+public interface ScoreDirectorFactoryProvider<Solution_, Score_ extends Score<Score_>> {
```
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/ScoreDirectorFactoryProvider.java:16]: Are we sure about this name?
Proposal A) ScoreDirectorFactoryProvider. Might be a bit close to "ConstraintProvider" and "ValueRangeProvider" with which I don't see the similarity.
Proposal B) ScoreDirectorFactoryBuilder. Doesn't really follow the "Builder" design pattern.
Proposal C) ScoreDirectorFactorySupplier. doesn't implement the Supplier function
Proposal D) ScoreDirectorFactoryProvisioner

I don't know. Your call. My preference order: D, A, B, C

File: optaplanner-quarkus-integration/optaplanner-quarkus/deployment/src/main/java/org/optaplanner/quarkus/deployment/OptaPlannerProcessor.java (hunk 6 above)
Comment (triceo) [reply optaplanner-quarkus-integration/optaplanner-quarkus/deployment/src/main/java/org/optaplanner/quarkus/deployment/OptaPlannerProcessor.java:113]: This comes straight from Quarkus documentation:
https://quarkus.io/guides/writing-extensions#service-files

**Comment by triceo** [comment]
The Sonar report got much worse with the last commit, as I have renamed several packages, and this brought unrelated classes to the scope of the analysis. THe original Sonar report was favorable and I intend to stick with that one.

**Pull request merged**